### PR TITLE
[14.0][ADD] stock_available_portal

### PR DESCRIPTION
**Pull request by geomer198**
This module extends the functionality of **Stock available to promise**.
It allows to show "Quantity available to promise" of your products in customer portal.
A new menu "Products" is added to the Portal view. So your portal users can see your current inventory.
You can configure which products are visible in portal and which users have access to these products.

### CODE REVIEW

File: stock_available_portal/models/product_product.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    _inherit = ["product.product", "portal.mixin"]
+
+    def _compute_access_url(self):
+        """Compute access url for portal view"""
+        for rec in self:
+            rec.access_url = "/my/products/{}".format(rec.id)
+
+    @api.model
+    def get_config_domain(self, key):
+        """Get domain by key from config"""
```
Comment (ivs-cetmix) [inline stock_available_portal/models/product_product.py]: Get domain using config key

File: stock_available_portal/models/res_config_settings.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    portal_visible_products_domain = fields.Char(
+        string="Portal Visible Products",
+        help="Domain which defines which products can be visible on portal. "
+        "By default no products.",
```
Comment (ivs-cetmix) [inline stock_available_portal/models/res_config_settings.py]: By défaut all product?

File: stock_available_portal/models/res_config_settings.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        config_parameter="stock_available_portal.portal_visible_products_domain",
+    )
+    portal_visible_users = fields.Char(
+        string="Visible to Users",
+        help="Domain which defines which users can see products. By default no-one.",
```
Comment (ivs-cetmix) [inline stock_available_portal/models/res_config_settings.py]: By default all users?

File: stock_available_portal/models/product_product.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        try:
+            return safe_eval(self.env["ir.config_parameter"].sudo().get_param(key, []))
+        except (TypeError, SyntaxError, NameError, ValueError):
+            return []
+
+    @api.model
+    def check_access_portal_products(self):
```
Comment (ivs-cetmix) [inline stock_available_portal/models/product_product.py]: check_product_portal_access

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
Looks almost good!
- Add /readme
- Fix minor spelling issues
- Try to improve test coverage to have codecov green
- Add CetmixGitDrone as mainliner in manifest

File: stock_available_portal/models/product_product.py (hunk 1 above)
Comment (geomer198) [reply stock_available_portal/models/product_product.py]: Resolved

File: stock_available_portal/models/res_config_settings.py (hunk 2 above)
Comment (geomer198) [reply stock_available_portal/models/res_config_settings.py]: Resolved

File: stock_available_portal/models/product_product.py (hunk 4 above)
Comment (geomer198) [reply stock_available_portal/models/product_product.py]: Resolved

File: stock_available_portal/models/res_config_settings.py (hunk 3 above)
Comment (geomer198) [reply stock_available_portal/models/res_config_settings.py]: Resolved

File: stock_available_portal/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+* Open **Settings** > **Inventory**. In the section **Stock available to promise** set domain for the access on "My Products"  in portal for users and set domain for the showing sets product on portal.
```
Comment (ivs-cetmix) [inline stock_available_portal/readme/USAGE.rst]: Open **Settings** > **Inventory**.
 In  **Stock available to promise** section there are domains for products and users. 
Product domain defines which products will be visible in portal.
User domain defines which users have access to products in portal.

File: stock_available_portal/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+This module adds the functionality to display **Products** in the portal.
+Access rights are defined in the **Settings**.
```
Comment (ivs-cetmix) [inline stock_available_portal/readme/DESCRIPTION.rst]: This module adds the functionality to display **Products** in  portal.
Configurator domains are used define which products are visible and which users have access to products

**Review by ivs-cetmix — APPROVED** [review]
All good.
Fix spelling, squash commits and move PR out of Draft

File: stock_available_portal/readme/USAGE.rst (hunk 5 above)
Comment (geomer198) [reply stock_available_portal/readme/USAGE.rst]: Resolved

File: stock_available_portal/readme/DESCRIPTION.rst (hunk 6 above)
Comment (geomer198) [reply stock_available_portal/readme/DESCRIPTION.rst]: Resolved

**Review by ivs-cetmix — APPROVED** [review]
Looks good!

**Comment by ivs-cetmix** [comment]
@OCA/logistics-maintainers would appreciate your review here

File: stock_available_portal/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+    "name": "Stock available to Portal",
```
Comment (ivs-cetmix) [inline stock_available_portal/__manifest__.py]: Stock Available In Portal

File: stock_available_portal/readme/DESCRIPTION.rst [hunk 8]
```diff
@@ -0,0 +1,3 @@
+This module adds the functionality to display **Products** in portal.
```
Comment (ivs-cetmix) [inline stock_available_portal/readme/DESCRIPTION.rst]: This module extends the functionality of **Stock available to promise**.
It allows to show  "Quantity available to promise" of your products in customer portal.
A new menu "Products" is added to the Portal view. So your portal users can see your current inventory.
You can configure which products are visible in portal and which users have access to these products.

File: stock_available_portal/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    "version": "14.0.1.0.0",
+    "author": "Cetmix, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
```
Comment (ivs-cetmix) [inline stock_available_portal/__manifest__.py:6]: Add "summary":
"Show product quantity available to promise in portal"

File: stock_available_portal/readme/USAGE.rst [hunk 10]
```diff
@@ -0,0 +1,5 @@
+Open **Settings** > **Inventory**.
```
Comment (ivs-cetmix) [inline stock_available_portal/readme/USAGE.rst]: **This text must be in CONFIGURE.rst** 

Open **General Settings** > **Inventory**.

 Configure the following filters in **Stock available to promise** section:

*  **Portal Visible Products** defines which products will be visible in portal
* **Visible to Users** defines which users have access to products in portal

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
1. Current USAGE.rst is actually CONFIGURE.rst Move its content to a proper location. See review comments before moving!
2. Add usage instructions to USAGE.rst
3.  Add module summary to PR description

**Comment by geomer198** [comment]
@isserver1 Resolved comments. Please approve changes.

**Comment by rousseldenis** [comment]
@geomer198 Shouldn't this be split with a base module that defines a product portal and then, this one that adds the available quantities ?

**Comment by francesco-ooops** [comment]
@geomer198 @isserver1 what if you have the need to display a different set of products to different portal users?

could it be an idea to extend https://github.com/OCA/product-attribute/tree/14.0/product_assortment to portal?

**Comment by ivs-cetmix** [comment]
> @geomer198 Shouldn't this be split with a base module that defines a product portal and then, this one that adds the available quantities ?

I might be wrong but showing products without availability is pointless.
This is why this module is built on top of "Quantity available to promise"

**Comment by ivs-cetmix** [comment]
> @geomer198 @isserver1 what if you have the need to display a different set of products to different portal users?
> 
> could it be an idea to extend https://github.com/OCA/product-attribute/tree/14.0/product_assortment to portal?

You will have to develop a bridge module for that

**Review by ivs-cetmix — APPROVED** [review]
LGTM!

**Review by Bearnard21 — APPROVED** [review]
Functional test passed:
Product filters tested
User filters tested

**Review by GabbasovDinar — APPROVED** [review]
Code LGTM

**Pull request merged**